docs/internals/optimizer.rst

```diff
@@ -1222,13 +1281,52 @@ code than more efficient code. In some cases, though, inlining a function
 can have positive effects on subsequent optimizer steps. This is the case
 if one of the function arguments is a constant, for example.
 
-During inlining, a heuristic is used to tell if the function call
-should be inlined or not.
-The current heuristic does not inline into "large" functions unless
-the called function is tiny. Functions that are only used once
-are inlined, as well as medium-sized functions, while function
-calls with constant arguments allow slightly larger functions.
+The transform changes code of the form
+
+.. code-block:: none
+
+    function f(a, b) -> c { /* ... */ }
+    let z := f(x, y)
+
+into
+
+.. code-block:: none
+
+    function f(a, b) -> c { /* ... */ }
+    let f_b := y
+    let f_a := x
+    let f_c
+    /* ... code of f, with replacements: a -> f_a, b -> f_b, c -> f_c */
+    let z := f_c
+
+During inlining, a heuristic is used to tell if the function call should be inlined or not.
+
+A function call will never be inlined if:
+
+- The function definition contains the ``leave`` opcode.
+- The function is large and inlining is not *aggressive* (details below).
+- The function definition contains any direct, recursive calls to itself or the call being inlined is recursive.
+  Note that indirect recursive calls (going through other functions) do not disqualify the function from inlining.
+- Argument expressions passed into the call can potentially have side effects.
+  This check is very broad, rejecting anything that is not simply a literal or an identifier, since
+  the condition is easy to satisfy by running ExpressionSplitter beforehand.
+
+Otherwise, if any of the following conditions is met, the call is inlined unconditionally:
+
+- The function is tiny.
+- The function is only ever called once.
+
+In the remaining cases the decision to inline or not depends on the size of the function.
+The size used here is not the actual bytecode size (which is not known to the optimizer), but rather
+a score calculated at the Yul level, which approximates that size.
+Inlining is performed if the size does not exceed a predefined threshold, which is a little higher
+if the inlining is aggressive and yet higher when the call has some constant arguments.
 
+The aggressive inlining is performed when:
+
+- The :ref:`yul-memoryguard` is present.
+- The function is not recursive (not even indirectly).
+- The legacy Yul->EVM transform is not used (i.e. the target is not the ``homestead`` :ref:`EVM version <evm-version>`).
```

Comment:
This is a summary of what [`FullInliner::shallInline()`](https://github.com/ethereum/solidity/blob/v0.8.25/libyul/optimiser/FullInliner.cpp#L178-L238) does.

Reply:
I really don't think it is good to document things into this level of detail :-). There is little merit in it, it only leads to people *relying* on some behaviour we want to be able to change and it's a maintenance hassle.

Reply:
This is something I needed to know multiple times in the past and each time I just looked at the code of `shallInline()`, but it's nowhere as clear as having it spelled out in the docs and I never had time to dive into details so I actually never fully know how it really worked. For example, that the `m_noInlineFunctions` check is simply a check for functions using `leave`. Or how some indirect recursive calls may actually be subject to inlining (there are three separate checks for recursive calls and it's not obvious how they differ). This time I needed it again to figure out why a function was not being inlined in one of my snippets. It always felt like something that should have been documented *somewhere*.

I could put it the docstring or inside `FullInliner::shallInline()`, but I really don't see much problem with having it in the docs proper. It's our main resource to understand how the optimizer works, not just for users but also for us and that fully justifies some extra effort spent keeping it up to date.

At the same time I don't think the extra detail really hurts users. It does not go into tiny details of how the inliner is implemented. It just describes the heuristic, which is useful to know when you're optimizing your code. But even then it omits all the specifics that could easily change, like numbers or details of the metric. There are step descriptions that are much more detailed than that. So if you're worried about users relying on them - we already have that problem. I don't think it's a good reason to avoid documenting things. I think that we in general lean too hard on that and I think it's one of the reasons why the code is so hard to get into.

In any case, I feel very strongly that the inliner heuristic should be documented.